jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java

```diff
@@ -420,25 +433,58 @@ else if (filled == 0)
             }
             finally
             {
-                if (networkBuffer.isRetained() && !shutdown)
+                try (AutoLock ignore = lock.lock())
                 {
-                    holdBuffer(networkBuffer);
+                    if (networkBuffer.isRetained() && this.heldBuffer != RELEASE_MARKER && !shutdown)
```

Comment:
Need some javadoc and/or comments to explain the important aspects of these changes?

Reply:
I've added some comments explaining the mechanism.